text/0000-glimmer-component-signature.md

```diff
@@ -0,0 +1,361 @@
+  Args: { post: Post };
+  BlockParams: {
+    default: [postTitle: string, postAuthor: string, postBody: string];
+  };
+}
+```
+
+Block parameters may also depend on the types of args a component receives. For example, for a fancy table component:
+
+```hbs
+<FancyTable @items={{this.people}}>
+  <:header>
+    <td>Name</td>
+    <td>Age</td>
+  </:header>
+
+  <:row as |person|>
+    <td>{{person.name}}</td>
+    <td>{{person.age}}</td>
+  </:row>
+</FancyTable>
+```
+
+The backing class and signature might look something like:
+
+```ts
+export interface FancyTableSignature<T> {
+  Args: {
+    /** The items to be displayed in the fancy table, each corresponding to one row. */
+    items: Array<T>
+  };
+  BlockParams: {
+    /** Any header contents for the table, broken into cells. */
+    header: [];
+
+    /** Content to be rendered for each row, receiving the corresponding item and its index. */
+    row: [item: T, index: number];
+  };
+}
+
+export default class FancyTable<T> extends Component<FancyTableSignature<T>> {
+  // ...
+}
+```
+
+### `Element`
+
+The `Element` member of the signature declares what type of DOM element(s), if any, this component may be treated as encapsulating. That is, setting a non-`null` type for this member declares that this component may have HTML attributes applied to it, and that type reflects the type of DOM `Element` object any modifiers applied to the component will receive when they execute.
+
+For example, [`{{animated-if}}`] would set `Element: null`, as it emits no DOM content. Even if you invoked it with angle-bracket syntax, any attributes or modifiers you applied wouldn't go anywhere.
+
+On the other hand, [`<ResponsiveImage>`] would set `Element: HTMLImageElement`, as the element in its template that it ultimately spreads `...attributes` on to is an `<img>`.
+
+While it's not common, occasionally components might forward `...attributes` to different types of elements in their template:
+
+```hbs
+{{#if @destination}}
+  <a ...attributes href={{@destination}}>{{yield}}</a>
+{{else}}
+  <span ...attributes>{{yield}}</span>
+{{/if}}
+```
+
+For such cases, components can use a union type for their `Element`. In the case of the template above, the signature would have `Element: HTMLAnchorElement | HTMLSpanElement`.
+
+[`{{animated-if}}`]: https://ember-animation.github.io/ember-animated/docs/api/components/animated-if
+[`<ResponsiveImage>`]: https://github.com/kaliber5/ember-responsive-image#the-responsiveimage-component
+
+## How we teach this
+
+### Documentation
+
+TypeScript is [not (currently) officially supported by Ember](https://github.com/emberjs/rfcs/pull/724), and as such none of the framework guides or API documentation mention the existing `Args` type parameter today. The upshot of this is that, while the `Args` type parameter is well known by the portion of the Ember community that works with TypeScript, there's no official documentation to be updated.
```

Comment:
The official documentation should be updated with the more appropriate definitions for blocks, block params and yields. Anyway, definitions wouldn't have been necessary as part of this RFC.

Reply:
I think the [modern guides](https://guides.emberjs.com/release/components/block-content/) (also linked in the text) do a pretty decent job defining these terms (though many folks who've been using the framework long enough have probably never read the newer versions of the guides).

The point here, though, isn't the terminology at all: it's just that the `@glimmer/component` type parameter isn't included in any official documentation (and rightly shouldn't be, given the current status of Ember + TypeScript)